add-utxo.py fails when providing utxo:vout on command line

v0.8.0

    $ python add-utxo.py $txid:$vout
    input private key for [txid]:[vout], in WIF compressed format : [key]
    Traceback (most recent call last):
      File "add-utxo.py", line 245, in <module>
        main()
      File "add-utxo.py", line 242, in main
        add_ext_commitments(utxo_data)
      File "add-utxo.py", line 51, in add_ext_commitments
        add_external_commitments(ecs)
      File "[path]/jmclient/jmclient/podle.py", line 375, in add_external_commitments
        update_commitments(external_to_add=ecs)
      File "[path]/jmclient/jmclient/podle.py", line 398, in update_commitments
        write_to_podle_file(commitments, external)
      File "[path]/jmclient/jmclient/podle.py", line 330, in write_to_podle_file
        externalfmt = external_dict_to_file(external)
      File "[path]/jmclient/jmclient/podle.py", line 316, in external_dict_to_file
        return switch_external_dict_format(ed, utxo_to_utxostr, bintohex)
      File "[path]/jmclient/jmclient/podle.py", line 296, in switch_external_dict_format
        assert success, "invalid utxo format in external dict parsing."
    AssertionError: invalid utxo format in external dict parsing.

The problem is that `utxostr_to_utxo()` returns a tuple of an error bool and the utxo but the error code is being ignored in 3 places in `add-utxo.py`.

Something like the following fixes it:

    diff --git a/scripts/add-utxo.py b/scripts/add-utxo.py
    index 40a19d1..788e454 100755
    --- a/scripts/add-utxo.py
    +++ b/scripts/add-utxo.py
    @@ -196,7 +196,9 @@ def main():
                     u, priv = get_utxo_info(ul)
                     if not u:
                         quit(parser, "Failed to parse utxo info: " + str(ul))
    -                utxo_data.append((utxostr_to_utxo(u), priv))
    +                success, utxo = utxostr_to_utxo(u)
    +                assert success, "invalid utxo format"
    +                utxo_data.append((utxo, priv))
         elif options.in_json:
             if not os.path.isfile(options.in_json):
                 jmprint("File: " + options.in_json + " not found.", "error")
    @@ -208,7 +210,9 @@ def main():
                     jmprint("Failed to read json from " + options.in_json, "error")
                     sys.exit(EXIT_FAILURE)
             for u, pva in iteritems(utxo_json):
    -            utxo_data.append((utxostr_to_utxo(u), pva['privkey']))
    +            success, utxo = utxostr_to_utxo(u)
    +            assert success, "invalid utxo format"
    +            utxo_data.append((utxo, pva['privkey']))
         elif len(args) == 1:
             u = args[0]
             priv = input(
    @@ -216,7 +220,9 @@ def main():
             u, priv = get_utxo_info(','.join([u, priv]))
             if not u:
                 quit(parser, "Failed to parse utxo info: " + u)
    -        utxo_data.append((utxostr_to_utxo(u), priv))
    +        success, utxo = utxostr_to_utxo(u)
    +        assert success, "invalid utxo format"
    +        utxo_data.append((utxo, priv))
         else:
             quit(parser, 'Invalid syntax')
         if options.validate or options.vonly:

I don't know if you want to be checking the error status after all 3 of these calls, but either way only the 2nd part of the returned tuple should be used as the utxo.

--- a/jmclient/jmclient/commitment_utils.py
+++ b/jmclient/jmclient/commitment_utils.py
@@ -9,15 +9,17 @@ def quit(parser, errmsg): #pragma: no cover
     parser.error(errmsg)
     sys.exit(EXIT_FAILURE)
 
-def get_utxo_info(upriv):
+def get_utxo_info(upriv, utxo_binary=False):
     """Verify that the input string parses correctly as (utxo, priv)
-    and return that.
+    and return that. If `utxo_binary` is true, the first element of
+    that return tuple is the standard internal form
+    (txid-in-binary, index-as-int).
     """
     try:
         u, priv = upriv.split(',')
         u = u.strip()
         priv = priv.strip()
-        success, utxo = utxostr_to_utxo(u)
+        success, utxo_bin = utxostr_to_utxo(u)
         assert success, utxo
     except:
         #not sending data to stdout in case privkey info
@@ -30,7 +32,8 @@ def get_utxo_info(upriv):
     except:
         jmprint("failed to parse privkey, make sure it's WIF compressed format.", "error")
         raise
-    return u, priv
+    utxo_to_return = utxo_bin if utxo_binary else u
+    return utxo_to_return, priv
     
 def validate_utxo_data(utxo_datas, retrieve=False, utxo_address_type="p2wpkh"):
     """For each (utxo, privkey), first

--- a/scripts/add-utxo.py
+++ b/scripts/add-utxo.py
@@ -74,7 +74,8 @@ def main():
                     "BE CAREFUL about handling private keys! "
                     "Don't do this in insecure environments. "
                     
-                    "Also note this ONLY works for standard (p2pkh or p2sh-p2wpkh) utxos."
+                    "Also note this ONLY works for standard p2wpkh (native segwit) "
+                    "or p2sh-p2wpkh (nested segwit) utxos."
     )
     add_base_options(parser)
     parser.add_option(
@@ -193,10 +194,10 @@ def main():
         for ul in utxo_info:
             ul = ul.rstrip()
             if ul:
-                u, priv = get_utxo_info(ul)
+                u, priv = get_utxo_info(ul.decode("utf-8"), utxo_binary=True)
                 if not u:
                     quit(parser, "Failed to parse utxo info: " + str(ul))
-                utxo_data.append((utxostr_to_utxo(u), priv))
+                utxo_data.append((u, priv))
     elif options.in_json:
         if not os.path.isfile(options.in_json):
             jmprint("File: " + options.in_json + " not found.", "error")
@@ -208,15 +209,19 @@ def main():
                 jmprint("Failed to read json from " + options.in_json, "error")
                 sys.exit(EXIT_FAILURE)
         for u, pva in iteritems(utxo_json):
-            utxo_data.append((utxostr_to_utxo(u), pva['privkey']))
+            utxobin, priv = get_utxo_info(",".join([u, pva["privkey"]]),
+                                                   utxo_binary=True)
+            if not utxobin:
+                quit(parser, "Failed to load utxo from json: " + str(u))
+            utxo_data.append((utxobin, priv))
     elif len(args) == 1:
-        u = args[0]
+        ul = args[0]
         priv = input(
-            'input private key for ' + u + ', in WIF compressed format : ')
-        u, priv = get_utxo_info(','.join([u, priv]))
+            'input private key for ' + ul + ', in WIF compressed format : ')
+        u, priv = get_utxo_info(','.join([ul, priv]), utxo_binary=True)
         if not u:
-            quit(parser, "Failed to parse utxo info: " + u)
-        utxo_data.append((utxostr_to_utxo(u), priv))
+            quit(parser, "Failed to parse utxo info: " + ul)
+        utxo_data.append((u, priv))
     else:
         quit(parser, 'Invalid syntax')
     if options.validate or options.vonly:

